### ISSUE TITLE
Unable to resolve a generic service registered as a generic implementation

### ISSUE THREAD

**Issue by Cogitaria** [simpleinjector/SimpleInjector#302]
The TLDR is that constructors accepting an `IService<Foo>` are not resolved because the `IService<Foo>` mapping never gets resolved (see step c below). 

--------------------------

Details:

a) I have a generic interface:  

``` c#
interface IService<T> ...
```

b) I have a generic implementation for it : 

``` c#
class Service<T> : IService<T> ... 
```

c) I register the interface with the service as follows: 

``` c#
container.Register<IService<Foo>, Service<Foo>>(Lifestyle.Singleton);
```

d) I have a registered class whose constructor accepts the interface:

``` c#
class Client(IService<Foo> service) ...
```

e) The application fails because it cannot resolve Client but the inner exception (drilling down the stack trace as usual) indicates it cannot resolve IService<Foo>, which most definitely was registered as indicated in **c)** 

I'm not sure if this helps but I can work around the issue as follows: 

Wrap the unresolved generic type: 

``` c#
class Wrapper : IService<Foo> 
```

Change the Client constructor to accept a Wrapper versus `IService<Foo>` 

The Client now resolves Wrapper without even having to register the type.

**Comment by dotnetjunkie** [comment]
What is the exact exception message? What does the container say it can't resolve `IService<Foo>`?

**Comment by Cogitaria** [comment]
I've refactored the offending code away (local sandbox changes so cannot run that again) but was able to setup the generic interface to generic type registration that successfully resolved.  So I'm not able to reproduce except in the application which uses SimpleInjector with TopShelf and also NancyFx.  The stack trace (the inner most exception as you always get those chains of cannot resolve dependencies) simply said it was not able to resolve the IService<Foo> reference in a constructor that referenced precisely that.

**Comment by Cogitaria** [comment]
Here is the exception.  I'm unable to inject generic interfaces, i.e. can only use a subset of C# types, and it only breaks at run-time. 

> {"Unable to resolve type: Services.IFoo`1[[System.Collections.Generic.List`1[[System.String, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]], mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]]"}

Here is a model of the original exception, and this code as-is also causes the exact same issue. I'm injecting an `IFoo<List<string>>` but IFoo<...> will fail to resolve regardless of what T is.   This will break even with all three types (IFoo, Foo and Client) in the same namespace and same assembly. 
```
public interface IFoo<T> where T : class  {  void Bar(T tee);   }

public class Foo<T> : IFoo<T> where T : class    {  public void Bar(T tee)   {    }    }

public class Client { public Client(IFoo<string>) ... 

```
I'll keep experimenting with it for a bit, see if I can't get more information on the root cause.

**Comment by dotnetjunkie** [comment]
Please show your registrations.

**Comment by dotnetjunkie** [comment]
The exception you show doesn't originate from Simple Injector, since Simple Injector never throws exceptions with this exact message. This leads me to believe that you are resolving from a different container. This should be easy to spot when analyzing the stack trace.

**Comment by Cogitaria** [comment]
I see that in the stack trace now that you mention it.  We are registering with SimpleInjector which uses TinyIoc as its default container.  NancyFx allows you to specify the IoC container, which we do.  There are obviously issues with that given that it's not working in this case.  I'll dig into it and report back in case it comes up for somebody else.

**Comment by Cogitaria** [comment]
NancyFx continues incorporates SimpleInjector when specified as the IoC but it's clearly not using it exclusively.  Here we have an example of a valid registration in SimpleInjector that does not work with TinyIoC.  I'm guessing it's more likely a NancyFx bug than a bug in either container.

**Comment by dotnetjunkie** [comment]
How are you integrating Simple Injector with NancyFx?
